File: day 7/AOC7.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func evaluateLeftToRight(numbers []int, operators []string) int {
	result := numbers[0]
	for i, op := range operators {
		switch op {
		case "+":
			result += numbers[i+1]
		case "*":
			result *= numbers[i+1]
		case "||":
			// Concatenate digits
			result, _ = strconv.Atoi(strconv.Itoa(result) + strconv.Itoa(numbers[i+1]))
		}
	}
	return result
}

func generateOperatorCombinations(n int, ops []string) [][]string {
	if n == 0 {
		return [][]string{{}}
	}
	
	smaller := generateOperatorCombinations(n-1, ops)
	var result [][]string
	
	for _, combo := range smaller {
		for _, op := range ops {
			newCombo := make([]string, len(combo)+1)
			copy(newCombo, combo)
			newCombo[len(combo)] = op
			result = append(result, newCombo)
		}
	}
	
	return result
}

func solveCalibration(filePath string) int {
	file, err := os.Open(filePath)
	if err != nil {
		fmt.Printf("Error opening file: %v\n", err)
		return 0
	}
	defer file.Close()

	totalCalibrationResult := 0
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ":")
		if len(parts) != 2 {
			continue
		}

		target, _ := strconv.Atoi(strings.TrimSpace(parts[0]))
		numStrs := strings.Fields(strings.TrimSpace(parts[1]))
		
		var numbers []int
		for _, numStr := range numStrs {
			num, _ := strconv.Atoi(numStr)
			numbers = append(numbers, num)
		}

		numOperators := len(numbers) - 1
		operators := []string{"+", "*", "||"}
		combinations := generateOperatorCombinations(numOperators, operators)

		valid := false
		for _, ops := range combinations {
			if evaluateLeftToRight(numbers, ops) == target {
				valid = true
				break
			}
		}

		if valid {
			totalCalibrationResult += target
		}
	}

	return totalCalibrationResult
}

func main() {
	fileName := "pattern.txt"
	result := solveCalibration(fileName)
	fmt.Printf("Total Calibration Result (with concatenation): %d\n", result)
}

File: 2024/day 7/AOC7.py
import os
from itertools import product

def evaluate_left_to_right(numbers, operators):
    """Evaluate the expression left-to-right, including concatenation."""
    result = numbers[0]
    for i, op in enumerate(operators):
        if op == '+':
            result += numbers[i + 1]
        elif op == '*':
            result *= numbers[i + 1]
        elif op == '||':
            # Concatenate the digits as a number
            result = int(str(result) + str(numbers[i + 1]))
    return result

def solve_calibration(file_path):
    total_calibration_result = 0

    with open(file_path, "r") as file:
        puzzle_input = file.read()

    for line in puzzle_input.strip().split("\n"):
        target, nums = line.split(":")
        target = int(target.strip())
        numbers = list(map(int, nums.strip().split()))
        
        num_operators = len(numbers) - 1
        valid = False

        # Generate all possible operator combinations
        for operators in product("+-*||", repeat=num_operators):
            if evaluate_left_to_right(numbers, operators) == target:
                valid = True
                break
        
        if valid:
            total_calibration_result += target

    return total_calibration_result

if __name__ == "__main__":
    # Ensure the input file is in the same folder as this script
    file_name = "pattern.txt"
    file_path = os.path.join(os.path.dirname(__file__), file_name)

    result = solve_calibration(file_path)
    print("Total Calibration Result (with concatenation):", result)

